src/Background/Portal.vala

```diff
@@ -0,0 +1,171 @@
+        if (desktop_integration == null) {
+            throw new DBusError.FAILED ("No connection to compositor.");
+        }
+
+        var apps = desktop_integration.get_running_applications ();
+        var results = new HashTable<string, Variant> (null, null);
+        foreach (var app in apps) {
+            var app_id = app.app_id;
+            if (app_id.has_suffix (".desktop")) {
+                app_id = app_id.slice (0, app_id.last_index_of_char ('.'));
+            }
+
+            results[app_id] = ApplicationState.RUNNING; //FIXME: Don't hardcode: needs implementation on the gala side
+        }
+
+        return results;
+    }
+
+    public async void notify_background (
+        ObjectPath handle,
+        string app_id,
+        string name,
+        out uint32 response,
+        out HashTable<string, Variant> results
+    ) throws DBusError, IOError {
+        uint32 _response = 2;
+        var _results = new HashTable<string, Variant> (str_hash, str_equal);
+
+        var notification_request = new NotificationRequest ();
+
+        notification_request.response.connect ((result) => {
+            switch (result) {
+                case NotificationRequest.NotifyBackgroundResult.CANCELLED:
+                    _response = 1;
+                    break;
+                case NotificationRequest.NotifyBackgroundResult.FAILED:
+                    break;
+                default:
+                    _response = 0;
+                    _results["result"] = result;
+                    break;
+            }
+
+            notify_background.callback ();
+        });
+
+        uint register_id = 0;
+        try {
+            register_id = connection.register_object<NotificationRequest> (handle, notification_request);
+        } catch (Error e) {
+            warning ("Failed to export request object: %s", e.message);
+            throw new DBusError.OBJECT_PATH_IN_USE (e.message);
+        }
+
+        notification_request.send_notification (app_id, name);
+
+        yield;
+
+        connection.unregister_object (register_id);
+        response = _response;
+        results = _results;
+    }
+
+    private enum AutostartFlags {
+        AUTOSTART_FLAGS_NONE,
+        AUTOSTART_FLAGS_DBUS_ACTIVATABLE
+    }
+
+    public bool enable_autostart (
+        string app_id,
+        bool enable,
+        string[] commandline,
+        uint32 flags
+    ) throws DBusError, IOError {
+        /* If the portal request is made by a non-flatpak application app_id will most of the time be empty */
+        if (app_id.strip () == "") {
+            /* Usually we can then asume that the first commandline arg is the app_id */
+            if (commandline[0].strip () != "") {
+                app_id = commandline[0];
+            } else {
+                return false;
+            }
+        }
+
+        var path = Path.build_filename (Environment.get_user_config_dir (), "autostart", app_id + ".desktop");
```

Comment:
This is better using a variable instead of replacing the app_id parameter.

Also, maybe using the full command as the filename will result in less confilicts in the case of no app_id ?

```suggestion
        var filename = app_id;

        /* If the portal request is made by a non-flatpak application app_id will most of the time be empty
         * Use the commandline as a fallback for the autostart filename.
         */
        if (filename == "") {            
            filename = string.joinv("-", commandline[0]).replace ("--", "-");
        }

        var path = Path.build_filename (Environment.get_user_config_dir (), "autostart", filename + ".desktop");
```

Reply:
It might be slightly confusing but is probably safer 👍